clients/roscpp/src/libros/service_server_link.cpp

```diff
@@ -362,14 +368,32 @@ bool ServiceServerLink::call(const SerializedMessage& req, SerializedMessage& re
     processNextCall();
   }
 
+  bool interrupted = false;
   {
     boost::mutex::scoped_lock lock(info->finished_mutex_);
 
     while (!info->finished_)
     {
-      info->finished_condition_.wait(lock);
+      if (timeout > ros::WallDuration(0.0))
+      {
+        if (info->finished_condition_.wait_for(lock, boost::chrono::nanoseconds(timeout.toNSec())) == boost::cv_status::timeout)
+        {
+          ROS_WARN("Service [%s] call failed: no response for %fsec", service_name_.c_str(), timeout.toSec());
+          interrupted = true;
+          break;
+        }
+      }
+      else
+      {
+        info->finished_condition_.wait(lock);
+      }
     }
   }
+  // Clear calls is properly protected
+  if (interrupted)
+  {
+    this->clearCalls();
```

Comment:
It seems this clears/cancels all calls instead of only the current one...?

Reply:
Good catch